app/Http/Controllers/CodexController.php

```diff
@@ -0,0 +1,445 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use App\Libraries\PDF_TOC;
+use App\Models\Codex;
+use App\Models\CodexSong;
+use App\Models\CodexText;
+use App\Models\CodexTextType;
+use App\Models\SongCategory;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Redirect;
+use Session;
+
+define('FPDF_FONTPATH', '/var/www/html/resources/fonts');
+
+class CodexController extends Controller
+{
+    public function index()
+    {
+        $codices = Codex::orderBy('name')->get();
+        $textTypes = CodexTextType::with('texts')->withCount('texts')->get();
+        $songTypes = SongCategory::orderBy('name')->with('songs')->withCount('songs')->get();
+
+        return view('codex.index', ['codices' => $codices, 'textTypes' => $textTypes, 'songTypes' => $songTypes]);
+    }
+
+    public function addSong()
+    {
+        if (! SongCategory::count()) {
+            Session::flash('flash_message', 'You need to add a song category first!');
+
+            return Redirect::route('codex::index');
+        }
+        $categories = SongCategory::orderBy('name')->get();
+
+        return view('codex.song-edit', ['song' => null, 'textType' => null, 'categories' => $categories, 'myCategories' => []]);
+    }
+
+    public function editSong(int $id)
+    {
+        $song = CodexSong::findOrFail($id);
+        $categories = SongCategory::orderBy('name')->get();
+        $myCategories = $song->categories->pluck('id')->toArray();
+
+        return view('codex.song-edit', ['song' => $song, 'categories' => $categories, 'myCategories' => $myCategories]);
+    }
+
+    public function storeSong(Request $request)
+    {
+        $song = new CodexSong();
```

Comment:
You can leave out the $song parameter and create the song inside the saveSong function